subsys/usb/device_next/class/usbd_cdc_ncm.c

```diff
@@ -0,0 +1,1282 @@
+		.bEndpointAddress = 0x01,                                              \
+		.bmAttributes = USB_EP_TYPE_BULK,                                      \
+		.wMaxPacketSize = sys_cpu_to_le16(64U),                                \
+		.bInterval = 0,                                                        \
+	},                                                                             \
+	\
+	.if1_1_hs_in_ep = {                                                            \
+			.bLength = sizeof(struct usb_ep_descriptor),                   \
+			.bDescriptorType = USB_DESC_ENDPOINT,                          \
+			.bEndpointAddress = 0x82,                                      \
+			.bmAttributes = USB_EP_TYPE_BULK,                              \
+			.wMaxPacketSize = sys_cpu_to_le16(512U),                       \
+			.bInterval = 0,                                                \
+	},                                                                             \
+	\
+	.if1_1_hs_out_ep = {                                                           \
+			.bLength = sizeof(struct usb_ep_descriptor),                   \
+			.bDescriptorType = USB_DESC_ENDPOINT,                          \
+			.bEndpointAddress = 0x01,                                      \
+			.bmAttributes = USB_EP_TYPE_BULK,                              \
+			.wMaxPacketSize = sys_cpu_to_le16(512U),                       \
+			.bInterval = 0,                                                \
+	},                                                                             \
+	\
+	.nil_desc = {                                                                  \
+			.bLength = 0,                                                  \
+			.bDescriptorType = 0,                                          \
+	},                                                                             \
+	};                                                                             \
+	\
+	const static struct usb_desc_header *cdc_ncm_fs_desc_##n[] = {                 \
+			(struct usb_desc_header *) &cdc_ncm_desc_##n.iad,              \
+			(struct usb_desc_header *) &cdc_ncm_desc_##n.if0,              \
+			(struct usb_desc_header *) &cdc_ncm_desc_##n.if0_header,       \
+			(struct usb_desc_header *) &cdc_ncm_desc_##n.if0_union,        \
+			(struct usb_desc_header *) &cdc_ncm_desc_##n.if0_ncm,          \
+			(struct usb_desc_header *) &cdc_ncm_desc_##n.if0_netfun_ncm,   \
+			(struct usb_desc_header *) &cdc_ncm_desc_##n.if0_int_ep,       \
+			(struct usb_desc_header *) &cdc_ncm_desc_##n.if1_0,            \
+			(struct usb_desc_header *) &cdc_ncm_desc_##n.if1_1,            \
+			(struct usb_desc_header *) &cdc_ncm_desc_##n.if1_1_in_ep,      \
+			(struct usb_desc_header *) &cdc_ncm_desc_##n.if1_1_out_ep,     \
+			(struct usb_desc_header *) &cdc_ncm_desc_##n.nil_desc,         \
+	};                                                                             \
```

Comment:
The way descriptors are configured in other classes seems a bit different:
https://github.com/zephyrproject-rtos/zephyr/blob/95dde52b1e85b64bcb1d9ebb8c92dddb5bdc93e1/subsys/usb/device_next/class/usbd_msc.c#L86-L93
https://github.com/zephyrproject-rtos/zephyr/blob/c211cb347e0af0a4931e0e7af3d93577bcc7af8f/subsys/usb/device_next/class/usbd_cdc_acm.c#L61-L78
https://github.com/zephyrproject-rtos/zephyr/blob/95dde52b1e85b64bcb1d9ebb8c92dddb5bdc93e1/subsys/usb/device_next/class/bt_hci.c#L91-L108

But yet, at some other place in a recently added class:
https://github.com/zephyrproject-rtos/zephyr/blob/95dde52b1e85b64bcb1d9ebb8c92dddb5bdc93e1/subsys/usb/device_next/class/usbd_uac2_macros.h#L649-L651

So I am not sure...

Reply:
I took the ECM driver as a template.  So anything questionable comes from there ;-)

Reply:
Thank you! I asked for clarification on the [chat](https://discord.gg/zephyrproject)...

Reply:
Ah, I figured some of it out: A main struct is defined with all the descriptors together:

https://github.com/zephyrproject-rtos/zephyr/blob/95dde52b1e85b64bcb1d9ebb8c92dddb5bdc93e1/subsys/usb/device_next/class/usbd_cdc_ecm.c#L60-L79

And then, to be able to have a variant for FullSpeed and HighSpeed, these structs fields are referenced in two different descriptors sets, an array of pointers:

https://github.com/zephyrproject-rtos/zephyr/blob/95dde52b1e85b64bcb1d9ebb8c92dddb5bdc93e1/subsys/usb/device_next/class/usbd_cdc_ecm.c#L778-L804

This makes all sense now! So this looks like the way forward to have several sets of descriptors without re-declaring everything twice, only picking different lists.

This means the way you did, and that ECM did, looks very canonical, and forward compatible with USB3 should that be implemented.

Reply:
This is where this new syntax got introduced:
https://github.com/zephyrproject-rtos/zephyr/commit/42f7e1b97fc7217ae972540c3574cfcedb526941

You have quite the latest code there! Already supporting bleeding edge features such as dual-speed support.